content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -1,38 +1,408 @@
+architectures. Not all architectures have a flat address space directly
+referenced by numeric values, although that is common. Some are
+represented by a base address plus an offset, for example. Therefore,
+the representation for type Address is hidden from us, the clients.
+Consequently we cannot simply treat address values as numeric values.
+Don't worry, though. The operations we need are provided.
+
+Package `System` declares these comparison functions, for example:
+
+.. code-block:: ada
+
+   function "<"  (Left, Right : Address) return Boolean;
+   function "<=" (Left, Right : Address) return Boolean;
+   function ">"  (Left, Right : Address) return Boolean;
+   function ">=" (Left, Right : Address) return Boolean;
+   function "="  (Left, Right : Address) return Boolean;
+
+These functions are intrinsic, i.e., built-in, meaning that the compiler
+generates the code for them directly at the point of calls. There is no
+actual function body for any of them so there is no performance penalty.
+
+Any private type directly supports the equality function, and
+consequently the inequality function, as well as assignment. What we
+don't get here is address arithmetic, again because we don't have a
+compile-time view of the actual representation. That functionality is
+provided by package `System.Storage_Elements`, a child package we will
+cover later. We should say though, that the need for address arithmetic
+in Ada is rare, especially compared to C.
+
+Having type `Address` presented as a private type is not, strictly
+speaking, required by the language. Doing so is a good idea for the
+reasons given above, and is common among vendors. Not all vendors do,
+though.
+
+Note that `Address` is the type of the result of the query attribute
+`Address`.
+
+We mentioned potentially needing to swap bytes in networking
+communications software, due to the differences in the "endianness" of
+the machines communicating. That characteristic is specified in `System`
+as follows:
+
+.. code-block:: ada
+
+   type Bit_Order is (High_Order_First, Low_Order_First);
+   Default_Bit_Order : constant Bit_Order := implementation-defined;
+
+`High_Order_First` corresponds to "big-endian" and `Low_Order_First` to
+little-endian. Here is a real-word example of use, in which we retrieve arbitrarily-typed values from a given buffer starting at a given index:
+
+.. code-block:: ada
+
+   generic
+      type Retrieved is private;
+   procedure Retrieve_4_Bytes
+     (Value  : out Retrieved;
+      Buffer : Byte_Array;
+      Start  : Index)
+   with
+     Pre => Start in Buffer'Range and then
+            Start + 3 in Buffer'Range;
+
+   procedure Retrieve_4_Bytes
+     (Value  : out Retrieved;
+      Buffer : Byte_Array;
+      Start  : Index)
+   is
+      ...
+   begin
+      Value := ...;
+      if Default_Bit_Order /= High_Order_First then -- we're not on a Big Endian machine
+         Value := Byte_Swapped (Value);
+      end if;
+   end Retrieve_4_Bytes;
```

Comment:
This should be converted into an actual source-code block using the code widget. 

Our ultimate goal is to have as many complete examples as possible in the courses of the *learn* website. Even if the code isn't really useful, we want to give readers the opportunity to experiment with it. Therefore, `:code-block`s should be limited to very small, one-line examples — or references to standard Ada packages, as you had above.

You could start with this, for example:

```
.. code:: ada compile_button project=Courses.Intro_To_Embedded_Sys_Prog.Low_Level_Programming.Retrieve_Example
 
    package Retrieve_Example is

       type Byte is new Natural range 0 .. 255
         with Size => 8;
       
       subtype Index is Natural;

       type Byte_Array is array (Index range <>) of Byte;

       generic
          type Retrieved is private;
       procedure Retrieve_4_Bytes
         (Value  : out Retrieved;
          Buffer : Byte_Array;
          Start  : Index)
         with
           Pre => Start in Buffer'Range and then
           Start + 3 in Buffer'Range;

    end Retrieve_Example;

    with System; use System;

    package body Retrieve_Example is

       procedure Retrieve_4_Bytes
         (Value  : out Retrieved;
          Buffer : Byte_Array;
          Start  : Index)
       is
          function Byte_Swapped (Value : Retrieved) return Retrieved is
            (Value);   -- implementation is missing!
       begin
          --  Value := ...;
          
          if Default_Bit_Order /= High_Order_First then -- we're not on a Big Endian machine
             Value := Byte_Swapped (Value);
          end if;
       end Retrieve_4_Bytes;   

    end Retrieve_Example;
```

Note that it's OK to have missing implementation for parts of the code that aren't relevant. In the example above, we could simply indicate that the implementation of the `Byte_Swapped` function is missing — as its implementation isn't essential to understand the code.

Reply:
I agree with the goal, but not for absolutely every case. This is just a fragment, and should not be made complicated by the extra code.